README.md
# BASH Scripts

I've been writing script for quite a while, mostly for my own edification and amusement. Often a script will come into being as a solution to a particular problem or to automate a task. I started out with DOS batch files and more recently Bash scripts. I've dabbled in several programming languages, but I haven't done much actual programming since my college days. I'm not a scripting expert or guru by any means. I'm always learning, and I'm always looking for ways to improve my scripts. To paraphrase Paul Valéry, the French poet, "A script is never finished, only abandoned."

I have a git server on my home network where I keep my scripts and other source code, but I'm sharing certian scripts that I feel might be useful to others on my GitHub account.

I maintain a small home network of about a dozen desktops and laptops. Most of them are older machines, some more than 10 years old. I generally don't get rid of a computer if I can get some use out of it. All of them are running Debian or Ubuntu based operating systems (Debian, Linux Mint, and LMDE). Most of the scripts in this repository should run under other distributions, but may require a few modifications.

All the scripts are licensed under the Gnu Public Licence, v2.0. Feel free to download and use them if you find them useful. Modify them to fit your own use cases. I don't ask that you send me any changes you make, but if you find a way to make them better, I'd appreciate you sharing them with me.

### The Scripts
**Note:** Many of these scripts source `functionlib` which should be available in your ~/bin directory.
- **battery-check** - Displays charge and health status of a laptop battery. Pulls its data from files in /sys/class/power_supply/BAT?.
- **catless** - Displays a text file using either cat or less, depending on the size of the file and the number of lines in the current display.
- **check-battery** - Displays charge and health status of a laptop battery using the upower utility.
- **check-updates** - Check for updates using apt-get or nala. With apt-get, provides a menu to install updates, clean apt cache, remove orphaned apps, or defer the update. If flatpak is installed, updates flatpak apps.
- **cinnkeys** - Creates a backup file for Cinnamon keybindings and restores keybindings from the backup.
- **functionlib** - A library of useful functions that can be sourced by a script.
- **install-vrtmgr** - Installs virt-manager, KVM/QEMU, and associated utilities.
- **ip-info** - Displays IP address information for wired and wireless connections to include MAC addresses, DNS addresses, and default gateway.
- **locale-fix.sh** - changes locale-gen to store locales in individual locale direectories instead of a single archive file.
- **pwfeedback** - Enables/disables password feedback when typing password at the comand line.
- **rename-host** - Rename the system's hostname, based on a set of rules.
- **ren-ext** - Rename capitalized, mixed capitalization, and non-standard file extensions.
- **ren-space** - Replaces spaces in a file name with underscores.
- **show-cron** - Displays anacron jobs and crontabs for users and root.
- **set-resume** - Sets the RESUME environment variable with the UUID of the swap partition.
- **system-info** - Displays system information in an easy-to-read format.
- **system-status** - Displays a snapshot of CPU, memory, & disk usage, as well as temperatures and uptime.
- **upper2lower** - Changes filenames to all lowercase.
- **verify-iso** - Verifies checksums for downloaded ISO files.
- **write-iso** - Writes ISO files to USB drive using `dd`. Lists available ISO files in the current directory and attached USB drives.

### Feedback:

Feel free to contact me with comments and suggestions. I can be reached through my blog, Twitter, and email.

* [GitHub](https://github.com/RickRomig/bashscripts)
* [Rick's Tech Stuff](https://ricktech.wordpress.com)
* [Twitter (@ludditegeek)](https://twitter.com/ludditegeek)
* Email: [rick.romig@gmail.com](mailto:rick.romig@gmail.com) or [rick.romig@mymetronet.net](mailto:rick.romig@mymentronet.net)

Rick Romig (*The Luddite Geek*)

19 October 2023

### DISCLAIMER

THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL I BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS AND SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


changelog.md
# Changelog for bashscripts

### 4 February 2021

**pwfeedback.sh 1.0.1**

- Incorporated variables for filenames.

- Modified line to set permissions for `/etc/sudoers.d/0pwfeedback` to read only for owner and group (root:root).

**system-info 2.1.8**

- Added code to display the type of memory in the system (DDR2, DDR3, DDR4, etc).

- Changed code to display the amount of physical memory as an integer amount (8 GB, 16 GB, etc).

- Added code to display the number of CPU threads.

- Added check to install `upower` utility if it's not already installed (needed for battery informaton).

- Changed the `upower` command to the 

### 3 February 2021

**pwfeedback.sh 1.0.0**

- Added to repository and updated README.md.

### 21 January 2021

**battery-check 0.2.8**

- Modified heredoc in `version()` function for better readabilty and format.

- Combined conditional statements in `battery_status()` and `battery_capacity()` functions to use double brackets.

- Changed the conditional statement at the beginning of the "Execution" section from double to single brackets and replaced the `<` with `-gt`.

- Fixed case statements by to include `;;` at the end of each case.

**check-battery 0.1.11**

- Modified heredoc in `version()` function for better readabilty and format.

- Combined conditional statements in the `warning_level()` function to use double brackets.

- Changed the conditional statement at the beginning of the "Execution" section from double to single brackets and replaced the `<` with `-gt`.

- Fixed case statements by to include `;;` at the end of each case.

- In the `ac_status()` function, placed the seconde upower command in double quotes to prevent word splitting.

**clean-bin 2.0.8**

- Removed unused variable `CURDIR`.

**check-updates 1.3.1**

- In `install_updates()`, changed `apt-get full-upgrade` to `apt-get dist-upgrade`.

- Added code to display how many packages can be upgraded.

### 27 December 2020

**set-resume-var 1.1.3**

- Replaced sample update-initramfs message with a function containing a heredoc.

### 25 December 2020

**battery-check 0.2.7**

- Modified `ac_status` function. Returns 0 if not connected to AC power, 1 if connected to AC power, and 2 if AC power status cannot be determined.

- Displays AC status based on status returned by `ac_status`.

- Checks for arguments other than `-h`, `--help`, or `--version`. Terminates the script if an invalid argument is passed.

**check-battery 0.1.10**

- Displays AC status based on status returned by `ac_status`.

- Checks for arguments other than `-h`, `--help`, or `--version`. Terminates the script if an invalid argument is passed.

### 24 December 2020

**rename-host 2.1.3**

- Added GNU GPL v2.0 to header.

- Removed instuction to reboot. All you should need to do to check the changes is to open a new terminal or a terminal tab.

**check-updates 1.3.0**

- Removed elasped time function call from `all_done` and `defer_updates` functions and placed it at the end of the script.

- Added GNU GPL v2.0 to header.

- In the `select` do loop, replaced `$OPT` with `$REPLY` in the `case` statement.

- Updated the options array for the `select` command.

### 26 September 2020

**ip-info 2.2.0**

- Changed the  `WIFI_SIG` variable to include `dBm` after the numerical value.

- Removed the check for the WIFI device model name. If `WIFI_SIG` is empty, the strong "Not available" will be assigned and displayed.
  
  - The Broadcom BCM4313 wireless card which did work with Ubuntu 18.04, did work when I reloaded the system with Debian 10. This may have been due to a different versions of  `iw` rather than the device itself.

### 24 September 2020

**battery-check 0.2.6 & check-battery 0.1.9**

- Added functions to display help and version information

### 21 September 2020

**battery-check 0.2.5 & check-battery 0.1.8**

- Added check to determine if the system is a laptop.

- Added check to determine if the laptop is connected to AC power.

### 14 September 2020

**set-resume-var 1.1.2**

- In the `awk` command that extracts the swap partition's UUID, changed `/swap/`to `TYPE="swap"`to clarify the search term.

- In testing the `SWAP_UUID` assigment, noted that Debian, LMDE, antiX, MX, and Linux Mint did not require `sudo` to run `/sbin/blkid`. Ubuntu 18.04 did require elevated privileges.

### 14 August 2020

**rename-host 2.1.2**

- Updated hostname rules to improve display.

- Backup copies of `hostname` and `hosts` files are maintained.

### 09 August 2020

**battery-check 0.2.4**

- Combined check for `charge_full*` and `energy_full*` into one if statement.

- Moved `BAT_HEALTH` variable assignment to be included with the CAP_PERCENT variable assignment in the 'else' portion of the if statement.

**check-battery 0.1.7**

- Checks to see if `upower` is installed and exits with an error message if it is not.

### 08 August 2020

**battery-check 0.2.3**

- Changed the order of the battery capacity display, showing the actual capacity before the design capacity, more closely resembling the `upower` output.

### 07 August 2020

**check-battery 0.1.5**

- Renamed from `bat-health` to `check-battery`.

- Added checks for case when the battery's actual capacity or design capacity is zero (battery is completely ded and unusable).

- Modified the execution section to use a single `if-then`statemenmt.

### 30 July 2020

**bat-health 0.1.1**

- Rename `bat-status` to `bat-health`.

- Rewrote most of the script to use `upower` to obtain information on battery charge and capacity rather than obtaining data from individual files under `/sys/class/power_supply/BAT` and performing calculations on the data.

### 04 July 2020

**rename-host 2.1.1**

- Added default case to the case statement that checks the new hostname against the rules.

### 03 July 2020

**rename-host 2.1.0**

- Script now  reads user input for the new hostname rather than passing it as a command line argument.


pwfeedback.sh
#!/bin/bash
##########################################################################
# Script Name  : pwfeedback.sh
# Description  : Enables/disables display of asterisks when entering password on CLI.
# Dependencies : none
# Arguments    : none
# Author       : Richard B. Romig, 27 Sep 2020
# Email        : rick.romig@gmail.com
# Comments     :
##########################################################################
# Copyright (C) 2021, Richard B. Romig
# Email: rick.romig@gmail.com
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program. If not, see <https://www.gnu.org/licenses/>.

# shellcheck disable=SC2014
# shellcheck disable=SC2034

## Variables ##

readonly SCRIPT=$(basename "$0")
readonly VERSION="1.0.1"
readonly UPDATED="04 Feb 2021"
readonly PWFB=$(sudo grep -wq 'pwfeedback' /etc/sudoers)
readonly PWFB_FILE="/etc/sudoers.d/0pwfeedback"

## Functions ##

enable_pwfeedback() {
  PWFB="$1"
  if [ -z "$PWFB" ] && [ ! -f "$PWFB_FILE" ]; then
    sudo sh -c 'echo "Defaults pwfeedback" > /etc/sudoers.d/0pwfeedback'
    sudo chmod 440 "$PWFB_FILE"
    echo "Password feedback enabled by $PWFB_FILE"
  else
    echo -n "Password feedback is already enabled by "
    [ -f "$PWFB_FILE" ] && echo "$PWFB_FILE" || echo "/etc/sudoers"
  fi
}

disable_pwfeedback() {
  PWFB="$1"
  if [ -f "$PWFB_FILE" ]; then
    sudo rm -f "$PWFB_FILE"
    echo "Password feeback is disabled. $PWFB_FILE has been deleted."
  elif [ -n "$PWFB" ]; then
    echo "Password feedback is enabled in /etc/sudoers."
    echo "Run the command 'sudo visudo' to edit the sudoers file and remove 'pwfeedback'."
  else
    echo "Password feedback is not enabled."
  fi
}

## Execution ##

sudo ls > /dev/null 2>&1
clear
echo "$SCRIPT v$VERSION (Updated: $UPDATED)"
echo -e "Enable/Disable password feedback in the terminal.\n"
OPTIONS=("Enable password feedback" "Disable password feedback" "Take no action")
PS3="Option: "
select OPT in "${OPTIONS[@]}"; do
  case "$REPLY" in
    1 ) enable_pwfeedback "$PWFB"; break ;;
    2 ) disable_pwfeedback "$PWFB"; break ;;
    3 ) echo "Exiting without making changes."; break ;;
    * ) echo -e "\e[91mInvalid choice.\e[0m Try again." ;;
  esac
done

exit
